dot/types/grandpa_test.go

```diff
@@ -7,55 +7,161 @@ import (
+	secondVote := GrandpaVote{
+		Hash:   common.MustHexToHash("0xc314327941fdd924bc67fd72651c40aececd485ca3e878c21e02abb40feae5bd"),
+		Number: uint32(10),
+	}
+
+	firstSignatureArray := mustHexTo64BArray(t, "0xd7292caacc62504365f179892a7399f233944bf261f8a3f66260f70e0016f2d"+
+		"b63922726b015c82dc7131f4730fbec61f71672a571453e51029bfb469070900f")
+
+	secondSignatureArray := mustHexTo64BArray(t, "0xb3c408b74905dfedfffa66f99f16fe8b938fd8df76a92225228a1ca07523"+
+		"0b99a2d9e173c561952e1e378b701915ca188d2c832ef92a3fab8e455f32570c0807")
+
+	var authorityID [32]byte
+	copy(authorityID[:], identityPubKey.Encode())
+
+	grandpaEquivocation := GrandpaEquivocation{
+		RoundNumber:     1,
+		ID:              authorityID,
+		FirstVote:       firstVote,
+		FirstSignature:  firstSignatureArray,
+		SecondVote:      secondVote,
+		SecondSignature: secondSignatureArray,
+	}
+	preVoteEquivocation := PreVoteEquivocation(grandpaEquivocation)
+	equivocationVote := NewGrandpaEquivocation()
+	err := equivocationVote.Set(preVoteEquivocation)
+	require.NoError(t, err)
+
+	equivocationProof := GrandpaEquivocationProof{
+		SetID:        1,
+		Equivocation: *equivocationVote,
+	}
+
+	actualEncoding := scale.MustMarshal(equivocationProof)
+	require.Equal(t, expectedEncoding, actualEncoding)
+}
+
+func Test_PreVoteEquivocation_ScaleCodec(t *testing.T) {
+	t.Parallel()
+	firstVote := GrandpaVote{
 		Hash:   common.Hash{0xa, 0xb, 0xc, 0xd},
 		Number: 999,
 	}
+	secondVote := GrandpaVote{
+		Hash:   common.Hash{0xd, 0xc, 0xb, 0xa},
+		Number: 999,
+	}
+	signature := [64]byte{1, 2, 3, 4}
+	keypair, err := ed25519.GenerateKeypair()
+	require.NoError(t, err)
+
+	var authorityID [32]byte
+	copy(authorityID[:], keypair.Public().Encode())
+
+	equivocation := GrandpaEquivocation{
+		RoundNumber:     0,
+		ID:              authorityID,
+		FirstVote:       firstVote,
+		FirstSignature:  signature,
+		SecondVote:      secondVote,
+		SecondSignature: signature,
+	}
 
-	enc, err := scale.Marshal(testVote)
+	equivPreVote := PreVoteEquivocation(equivocation)
+	equivVote := NewGrandpaEquivocation()
+	err = equivVote.Set(equivPreVote)
 	require.NoError(t, err)
-	require.Equal(t, exp, enc)
+	encoding := scale.MustMarshal(*equivVote)
 
-	dec := GrandpaVote{}
-	err = scale.Unmarshal(enc, &dec)
+	grandpaEquivocation := NewGrandpaEquivocation()
+	err = scale.Unmarshal(encoding, grandpaEquivocation)
 	require.NoError(t, err)
-	require.Equal(t, testVote, dec)
+	require.Equal(t, equivVote, grandpaEquivocation)
+}
 
+func Test_GrandpaVote(t *testing.T) {
+	t.Parallel()
+	expectedEncoding := common.MustHexToBytes("0x0a0b0c0d00000000000000000000000000000000000000000000000000000000e7030000")
+	vote := GrandpaVote{
+		Hash:   common.Hash{0xa, 0xb, 0xc, 0xd},
+		Number: 999,
+	}
+
+	encoding := scale.MustMarshal(vote)
+	require.Equal(t, expectedEncoding, encoding)
+
+	grandpaVote := GrandpaVote{}
+	err := scale.Unmarshal(encoding, &grandpaVote)
+	require.NoError(t, err)
+	require.Equal(t, vote, grandpaVote)
```

Comment:
there isn't much value in such test that use our scale package for both marshal and unmarshal. 

Was about expectedEncoding created using some rust equivalent? Add info to get expectedEncoding in a comment

Reply:
I think there is still value if the expectedEncoding is derived from rust scale encodings. It allows us to assert that both the encoded and raw data are correct. I will like how the expected encoding was generated